### PR TITLE
Fix S1226 FP: Parameters wrongly considered as unread before reassignment

### LINKED ISSUE #2555: Fix S1226 FP: Parameters wrongly considered as unread before reassignment
### Description

According to [RSPEC-1226](https://rules.sonarsource.com/csharp/RSPEC-1226) 

> parameters should be [...] at least read before reassignment

Is null-checking or passing a parameter to a method like `string.IsNullOrWhiteSpace(text)` not **reading** the parameter?

### Repro steps

    public class Class1
    {
        public void Foo(string text)
        {
            if (string.IsNullOrWhiteSpace(text))
            {
                text = "(empty)";
            }

            Bar(text);
        }

        private void Bar(string text)
        {
            // Do something
        }
    }

### Expected behavior

The variable `text` has been used before it was assigned. So this code should be compliant.

### Actual behavior

The code is marked as non-compliant.

### Known workarounds

Use the parameter to assign a local variable.

### Related information

* SonarC# Version: 7.15
* Visual Studio Version: 16.2.0 (VS2019)

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #2555

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ParameterAssignedTo.vb [hunk 1]
```diff
@@ -1,8 +1,11 @@
-﻿Imports System
+﻿
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ParameterAssignedTo.vb]: _minor_ additional empty line at the start of the file

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs [hunk 2]
```diff
@@ -54,18 +72,52 @@ protected sealed override void Initialize(SonarAnalysisContext context)
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(SupportedDiagnostics[0], left.GetLocation(), left.ToString()));
                     }
                 },
-                SyntaxKindsOfInterest.ToArray());
+                SyntaxKindOfInterest);
         }
 
+        private bool IsReadBefore(SemanticModel semanticModel, ISymbol parameterSymbol, TAssignmentStatementSyntax assignment)
+        {
+            //FIXME: REMOVE DEBUG
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs]: There is still a `FIXME` to be removed :)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs [hunk 3]
```diff
@@ -54,18 +72,52 @@ protected sealed override void Initialize(SonarAnalysisContext context)
...
+        }
 
-        protected abstract bool IsReadBefore(SemanticModel semanticModel, ISymbol parameterSymbol, TAssignmentStatementSyntax assignment);
+        /// <summary>
+        /// Returns all nodes before the specified statement to the declaration of variable/parameter given by stopLocation.
+        /// This method recursively traverses all parent blocks of the provided statement.
+        /// </summary>
+        private static IEnumerable<SyntaxNode> GetPreviousNodes(Location stopLocation, SyntaxNode statement) 
+        {
+            if (statement == null || (stopLocation != null && statement.GetLocation().SourceSpan.IntersectsWith(stopLocation.SourceSpan)))   //Method declaration or Catch variable declaration, stop here and do not include this statement
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs]: This null check: `stopLocation != null` is unnecessary here, as it is already checked for null before calling this method

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs [hunk 4]
```diff
@@ -54,18 +72,52 @@ protected sealed override void Initialize(SonarAnalysisContext context)
...
+            {
+                return new SyntaxNode[] { };
+            }
+            var previousNodes = statement.Parent.ChildNodes()
+                .TakeWhile(x => x != statement)     //Take all from beginning, including "catch ex" on the way, down to current statement
+                .Reverse()                          //Reverse in order to keep the tail
+                .TakeWhile(x => stopLocation == null || !x.GetLocation().SourceSpan.IntersectsWith(stopLocation.SourceSpan))    //Keep the tail until "catch ex" or "int i" is found
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs]: This null check: `stopLocation == null` is unnecessary here, as it is already checked for null before calling this method

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ParameterAssignedTo.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@ public static class ParameterAssignedToStatic
         static void f7(this int a)
         {
             a = 42; // Noncompliant
-//          ^
+                    //          ^
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ParameterAssignedTo.cs]: I can't wait for the new IssueLocationCollector :))

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ParameterAssignedTo.vb (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ParameterAssignedTo.vb]: That's intentional VB convention/habit to separate all blocks, this situation included. There used to be 
`Option Strict/Explicit/Compare/Infer On/Off`
as first line of the file followed by an empty line and imports.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs]: This review is exactly the reason why I introduced FIXME concept :))

**Pull request merged**